### ISSUE TITLE
Define Logic

### ISSUE THREAD

**Issue by flowlo** [coduno/api#42]
We [currently ](https://github.com/coduno/api/blob/cc60e1015851fe4e25346218cfeaff15bb14b4f7/model/logic.go) lack a definition of `model.Logic`.

I thought of a way on how we could add modular logic blocks to our API without sacrificing much. The idea comes from `crypto`.

 * An exported function `RegisterLogic`, analogous to [`crypto.RegisterHash`](https://golang.org/pkg/crypto/#RegisterHash) in some package (like `github.com/coduno/api/logic`) in this repository. It will just build a custom "function table".
 * A predefined interface, analogous to ([`hash.Hash`](https://golang.org/pkg/hash/#Hash)) that every `Logic` has to offer.
 * Every implementation of logic calls `logic.Register` in it's `init()`. Those implementations can live inside this repository, or in private ones.
 * `Logic` can be instantiated as needed.
 * If some `Logic` is not linked, we `panic`.

This way, we can write our logic blocks in Go witth nearly no cuts in performance, and can keep them private if needed.

One drawback certainly is that we cannot magically add new `Logc` or types of logic at runtime.

Comments requested.

**Comment by victorbalan** [comment]
I think this is ok for now and for the future too because I don`t think we will have too many different types of logic. Maybe one will be multiply with weights without taking in account the time and another one that takes in account the time. What would the `Logic` interface have? I'm thinking about different functions for computing every skill, right?

**Comment by flowlo** [comment]
We currently have `Logic` in `Challenge` and `Task`. 

 * Will they consume the same arguments?

For example to compute the result of a single `Task`, `Logic` would have to access all previous submissions:

```go
func Logic interface {
	Compute(task, user *datastore.Key, submissions ...*datastore.Key) model.Skills
}
```

Or to compute the result of a challenge:

```go
func Logic interface {
	Compute(result *datastore.Key) model.Skills
}
```

Of course this can be varied. Please propose some interface declaration or comment on mine.

**Comment by victorbalan** [comment]
I think this is the best solution:

**Comment by victorbalan** [comment]
This is ok
```go
func Logic interface {
	Compute(key *datastore.Key) model.Skills
}
```

The key can be either task or result key. We can do a switch on the kind so we have just one Logic interface. For result it will just make the average of the tasks skill values and for the task we`ll figure out some formulas to calculate the skills. Would this make sense?

**Comment by flowlo** [comment]
If it's a task key, how would you determine for which user the result is to be generated?

**Comment by victorbalan** [comment]
you`re right..that kind of means we need to have two different Logic interfaces.

**Comment by flowlo** [comment]
So we would end up with something like this:

```go
// Resulter is a piece of code that can aggregate the outcome of some tasks and
// form a Result. It will populate the result passed.
type Resulter interface {
    Result(result *datastore.Key)
}

// Tasker is a piece of code that can reason about the submissions made for
// the passed task. It will return the results for the passed task and user.
type Tasker interface {
    Result(task, user *datastore.Key) model.Skills
}
```

A `Resulter` would then be responsible to call all the `Tasker`s and save the computed results. The only thing I am a bit worried about: We currently have no way to save more detailed results on a per-task basis.
